contracts/funding/FundingFactory.sol

```diff
@@ -1,29 +1,92 @@
 
-pragma solidity ^0.5.0;
+pragma solidity ^0.6.0;
 
-import "@openzeppelin/upgrades/contracts/Initializable.sol";
+//import "@openzeppelin/upgrades/contracts/Initializable.sol";
 import "./Funding.sol";
-import "./chainlink/PoeChainlinkClient.sol";
+import "./FundingOracleClient.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/token/ERC721/ERC721.sol";
+import "@openzeppelin/contracts-ethereum-package/contracts/utils/Counters.sol";
 
-contract FundingFactory is Initializable {
+contract FundingFactory is Initializable,ERC721UpgradeSafe {
+    using Counters for Counters.Counter;
+    Counters.Counter private _tokenIds;
 
-  event FundingCreated(address indexed funding, address owner, address operator, address interestToken) ;
+    event FundingCreated( 
+        address indexed funding,
+        address owner,
+        address operator,
+        address interestToken
+    );
 
-  address public link;
+    address public link;
 
-  function initialize(
-    address _link
-  ) public initializer {
-    link = _link;
-  }
+    function initialize(address _link) public initializer {
+        link = _link;
+    }
 
-  function createPoeFunding(address _cToken, address _operator,
-      address _oracle, bytes32 _jobId, uint8 _level, uint256 ticketPrice) public  returns (address fundingAddress) {
-    FundingChainlinkClient oracle = new PoeChainlinkClient(link, _oracle, _jobId, _level);
-    Funding funding = new Funding(msg.sender, _cToken, _operator, address(oracle), ticketPrice);
+    function createFunding(
+        address _cToken,
+        address _operator,
+        address _oracle,
+        bytes32 _jobId,
+        uint8 _level,
+        uint256 _ticketPrice,
+        address _trustedForwarder
+    ) public returns (address fundingAddress) {
+        FundingOracleClient oracle = new FundingOracleClient(
+            _oracle,
+            _jobId,
+            _level,
+            link
+        );
+        Funding funding = new Funding(
+            msg.sender,
+            _cToken,
+            _operator,
+            address(oracle),
+            _ticketPrice
+          
+        );
+        funding.transferOwnership(msg.sender);
 
-    fundingAddress = address(funding);
-    oracle.transferOwnership(fundingAddress);
-    emit FundingCreated(fundingAddress, msg.sender, _operator, _cToken);
-  }
+        fundingAddress = address(funding);
+        oracle.transferOwnership(fundingAddress);
+        emit FundingCreated(fundingAddress, msg.sender, _operator, _cToken);
+    }
+
+    function createFundingWithNft(
+        address _nftToken,
+        address _operator,
+        address _oracle,
+        bytes32 _jobId,
+        uint8 _level,
+        uint256 _ticketPrice,
+        string memory _name,
```

Comment:
Please supply the ERC721 contract address as an argument, so we can use every type of ERC721. I have interesting scenarios for this in the futute